### PR TITLE
feat(testing): add __repr__ to testing.Result

### PR DESCRIPTION
**Pull request by RioAtHome**
# Summary of Changes
Added `__repr__` to `testing.Result`, where it shows the status code, content-type header and first 20 char.
Also added a simple test.

# Related Issues
Issue #2044 

# Pull Request Checklist

- [x] Applied changes to both WSGI and ASGI code paths and interfaces (where applicable).
- [x] Added **tests** for changed code.
- [x] Prefixed code comments with GitHub nick and an appropriate prefix.
- [x] Coding style is consistent with the rest of the framework.
- [x] Updated **documentation** for changed code.
    - [ ] Added docstrings for any new classes, functions, or modules.
    - [ ] Updated docstrings for any modifications to existing code.
    - [ ] Updated both WSGI and ASGI docs (where applicable).
    - [ ] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [ ] Updated all relevant supporting documentation files under `docs/`.
    - [ ] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [ ] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [x] Changes (and possible deprecations) have [towncrier](https://towncrier.readthedocs.io/en/actual-freaking-docs/index.html) news fragments under `docs/_newsfragments/`, with the file name format `{issue_number}.{fragment_type}.rst`. (Run `towncrier --draft` to ensure it renders correctly.)

*PR template inspired by the attrs project.*

### CODE REVIEW

File: falcon/testing/client.py [hunk 1]
```diff
@@ -347,6 +347,13 @@ def json(self) -> Optional[Union[dict, list, str, int, float, bool]]:
 
         return json_module.loads(self.text)
 
+    def __repr__(self):
+        body_text = '{}'.format(self.text[:20])
+
+        return '<Status: {}, Content-Type: {}, Text: {}>'.format(
+            self.status_code, self.headers['Content-Type'], body_text
```
Comment (vytas7) [inline falcon/testing/client.py:354]: What happens if the response payload is invalid UTF-8?
```python
>>> testing.Result([b'\xc3\x28'], falcon.HTTP_418, {})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/RioAtHome/falcon/falcon/testing/client.py", line 351, in __repr__
    body_text = '{}'.format(self.text[:20])
  File "/tmp/RioAtHome/falcon/falcon/testing/client.py", line 339, in text
    self._text = self.content.decode(encoding)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc3 in position 0: invalid continuation byte
```

It is IMHO better to prevent `__repr__` from raising almost at any cost, since it can be very confusing to debug (IIRC the popular `pytest` has some safeguard to handle even this scenario, but it ain't pretty).

File: tests/test_utils.py [hunk 2]
```diff
@@ -764,6 +764,13 @@ def test_status(self, app):
         result = client.simulate_get()
         assert result.status == falcon.HTTP_702
 
+    def test_repr_result(self, app):
```
Comment (vytas7) [inline tests/test_utils.py:767]: IMHO it would be nice to add more tests both for WSGI and ASGI flavours of the framework, potentially using `simulate_request()` and friends.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Thanks for getting the ball rolling, this is a great start.

I've provided some feedback inline.

File: falcon/testing/client.py (hunk 1 above)
Comment (RioAtHome) [reply falcon/testing/client.py:354]: How about making `__repr__` log the content of the payload instead of text?

[the remark this answers is not in the document]
Comment (RioAtHome) [reply falcon/testing/client.py]: i made some changes, made `__repr__` start with prefix `Result<...>`, and put an if statement to check whether content-type header is available or not, so nothing can get raised from `__repr__`.

File: falcon/testing/client.py [hunk 3]
```diff
@@ -347,6 +347,13 @@ def json(self) -> Optional[Union[dict, list, str, int, float, bool]]:
 
         return json_module.loads(self.text)
 
+    def __repr__(self):
+        content_type = ''
+        if 'Content-Type' in self.headers.keys():
```
Comment (vytas7) [inline falcon/testing/client.py:352]: Not that we care too much about the performance of the testing utils, but this is a fairly wasteful test for inclusion...
IMHO it is much easier to simply do `self.headers.get('Content-Type', '')`.

File: falcon/testing/client.py [hunk 4]
```diff
@@ -347,6 +347,13 @@ def json(self) -> Optional[Union[dict, list, str, int, float, bool]]:
...
+            content_type = self.headers['Content-Type']
+
+        return 'Result<{} {} {}>'.format(self.status, content_type, self.content[:20])
```
Comment (vytas7) [inline falcon/testing/client.py:355]: So if `content_type == ''`, would you get two spaces between status and content?

**Review by vytas7 — CHANGES_REQUESTED** [review]
This is shaping up!
Should we also add a towncrier newsfragment for this improvement?

File: falcon/testing/client.py (hunk 4 above)
Comment (RioAtHome) [reply falcon/testing/client.py:355]: Ahh, sorry about that.
Will correct in the next commit.

**Comment by RioAtHome** [comment]
> This is shaping up! Should we also add a towncrier newsfragment for this improvement?

towncrier newsfragment?
Sorry to be a bother, but any pointers on what this is?, and where to start if i want to add it?

**Comment by CaselIT** [comment]
> towncrier newsfragment?

It's just a few words that should appear on the changelog. See other newsfragments here https://github.com/falconry/falcon/tree/master/docs/_newsfragments and this section of the contributing file https://github.com/falconry/falcon/blob/master/CONTRIBUTING.md#changelog

File: falcon/testing/client.py (hunk 4 above)
Comment (RioAtHome) [reply falcon/testing/client.py:355]: First way that came to my mind to handle extra spaces is this:
```
    def __repr__(self):
        content_type = self.headers.get('Content-Type', '')
        repr_result = 'Result<{} {} {}>'.format(
            self.status, content_type, self.content[:20]
        )

        return " ".join(repr_result.split())
```
And i have a feeling this isn't the best way to go about it.

Or i can change the separator instead of spaces, to colon or maybe a comma.
Like this `Result<200 OK::b'hello'>`
What do you think?

**Comment by RioAtHome** [comment]
@CaselIT Thanks for the replying.
I like to point out that the link to towncrier in the changelog section of the contributing file does not work.

**Comment by vytas7** [comment]
Unfortunately towncrier isn't particularly well maintained; we're looking for alternatives, but I don't know when the change could actually happen.
Anyway, I'll update the link to https://towncrier.readthedocs.io/, that should be fixed as per https://github.com/twisted/towncrier/issues/367 (although it still says `17.08` in the title? :thinking:)

File: falcon/testing/client.py (hunk 4 above)
Comment (RioAtHome) [reply falcon/testing/client.py]: Changed the syntax to `Result<Status, Content-Type, Content>`.
Where now if no content type header was included, `__repr__` will give a prettier result imo.

File: falcon/testing/client.py (hunk 3 above)
Comment (RioAtHome) [reply falcon/testing/client.py]: i have edited it to `self.headers.get('Content-Type', '')`.

File: falcon/testing/client.py [hunk 5]
```diff
@@ -347,6 +347,11 @@ def json(self) -> Optional[Union[dict, list, str, int, float, bool]]:
 
         return json_module.loads(self.text)
 
+    def __repr__(self):
+        content_type = self.headers.get('Content-Type', '')
+
+        return 'Result<{}, {}, {}>'.format(self.status, content_type, self.content[:20])
```
Comment (myusko) [inline falcon/testing/client.py:353]: Since we are supporting CPython3.5+, I'd suggest using the `f-strings` instead of `str.format` method, does it make sense?

Comment (vytas7) [reply falcon/testing/client.py:353]: Probably not yet @myusko, there is a separate issue for dropping that support -- let's take one step at a time.

Comment (vytas7) [reply falcon/testing/client.py:353]: @myusko to make it clear, we do support CPython3.5+, but 3.5 does not support f-strings.
We will drop 3.5 & 3.6 support in https://github.com/falconry/falcon/issues/2049.

Comment (myusko) [reply falcon/testing/client.py:353]: @vytas7 Got it, tbh I've thought that `f-strings` was introduced in 3.5:) sorry, my bad!

**Comment by RioAtHome** [comment]
Hello, sorry to be a bother, but can i get a review on this pull request?

File: tests/test_utils.py [hunk 6]
```diff
@@ -764,6 +764,52 @@ def test_status(self, app):
...
+        )
+
+        assert str(result) == expected_result
+
+    def test_repr_without_content_type_header(self, asgi):
+        value = b'huh'
+        header = [('Not-content-type', '!!!')]
+        result = falcon.testing.Result([value], falcon.HTTP_200, header)
+
+        expected_result = 'Result<200 OK, , {}>'.format(value)
```
Comment (vytas7) [inline tests/test_utils.py:810]: I know I'm nitpicking, but OTOH a representation is about the esthetics :smiling_imp:
What if `Content-Type` or content are empty, the `__repr__` would look somewhat awkward -- as illustrated here.

Maybe we could skip the `Content-Type` if it is empty? Those commas are probably not needed either, although I'm open for suggestions there.

My suggestions could be
`Result<200 OK application/json b'{}'>`
`Result<201 Created>`
`Result<503 Service Unavailable b'Database error, dunno\n'>`

One can also ellipsize the text that was truncated, use lenient decoding and present data as string, the possibilities are endless :slightly_smiling_face:

File: docs/_newsfragments/2044.newandimporved.rst [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Added ``__repr__`` method to :class:`falcon.testing.client.Result`.
+Syntax of ``__repr__`` as follows ``Result<{status_code}, {Content-Type Header}, {Content}>``.
```
Comment (vytas7) [inline docs/_newsfragments/2044.newandimporved.rst:2]: Suggested edit:
> The form of `__repr__` is as follows: `...`, where the _content_ part will reflect up to 20 bytes of the result's content.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Yes, sorry for taking time @RioAtHome...
This PR is looking good, just a couple of tweaks and it's IMHO ready for merging. :shipit:

File: tests/test_utils.py (hunk 6 above)
Comment (CaselIT) [reply tests/test_utils.py:810]: this would be nice to have, but can probably be deferred until some later time

**Review by CaselIT — APPROVED** [review]

File: tests/test_utils.py (hunk 6 above)
Comment (RioAtHome) [reply tests/test_utils.py]: > this would be nice to have, but can probably be deferred until some later time

So sorry, i missed this message before implementing the suggested syntax, my latest commit implement the suggestions above... my apologies.

Concerning my latest commit that changes the syntax, i couldn't think of a way to implement the representation of `Result.content` without an if statement, any feedback will be much appreciated!

**Comment by RioAtHome** [comment]
my latest commit changes:
- `Result.__repr__` will not show any `None` values.
- `Result.__repr__` shows a total of 40 bytes of Result.content where it takes the first 20 and the last 20 and ellipsize the rest with `...`.
- made some few adjustments to `Result.__repr__` tests, mainly removed duplicate code (the get and post methods).

File: docs/_newsfragments/2043.bugfix.rst [hunk 8]
```diff
@@ -1,3 +0,0 @@
-Request attributes :attr:`~falcon.Request.forwarded_scheme` and
-:attr:`~falcon.Request.forwarded_host` now no longer raise an ``IndexError``
-while processing an invalid or empty ``Forwarded`` header.
```
Comment (CaselIT) [inline docs/_newsfragments/2043.bugfix.rst:3]: this seems unrelated

**Review by CaselIT — COMMENTED** [review]
it seems that an unrelated file was removed. looks ok for me otherwise

**Review by CaselIT — DISMISSED** [review]

**Review by vytas7 — DISMISSED** [review]
This LGTM now :+1: 
(I've taken the liberty of polishing the newsfragment somewhat.)
